File: packages/oxlint-plugin-react-doctor/src/plugin/rules/nextjs/nextjs-no-edge-og-runtime.ts
import { OG_IMAGE_FILE_PATTERN } from "../../constants/nextjs.js";
import { defineRule } from "../../utils/define-rule.js";
import { normalizeFilename } from "../../utils/normalize-filename.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { Rule } from "../../utils/rule.js";
import type { RuleContext } from "../../utils/rule-context.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";

export const nextjsNoEdgeOgRuntime = defineRule<Rule>({
  id: "nextjs-no-edge-og-runtime",
  title: "Edge runtime in OG image route",
  tags: ["test-noise"],
  requires: ["nextjs"],
  severity: "warn",
  recommendation:
    "Remove `export const runtime = 'edge'` from OG image files. The default Node.js runtime supports more fonts and APIs",
  create: (context: RuleContext) => {
    let isOgImageFile = false;

    return {
      Program() {
        const filename = normalizeFilename(context.filename ?? "");
        isOgImageFile = OG_IMAGE_FILE_PATTERN.test(filename);
      },
      ExportNamedDeclaration(node: EsTreeNodeOfType<"ExportNamedDeclaration">) {
        if (!isOgImageFile) return;

        const declaration = node.declaration;
        if (!isNodeOfType(declaration, "VariableDeclaration")) return;

        for (const declarator of declaration.declarations ?? []) {
          if (!isNodeOfType(declarator, "VariableDeclarator")) continue;
          if (!isNodeOfType(declarator.id, "Identifier")) continue;
          if (declarator.id.name !== "runtime") continue;

          const initValue = isNodeOfType(declarator.init, "Literal") ? declarator.init.value : null;

          if (initValue === "edge") {
            context.report({
              node,
              message:
                "Edge runtime limits OG image generation. Node.js runtime supports more fonts, filesystem access, and larger response sizes.",
            });
          }
        }
      },
    };
  },
});

File: native/oxlint/rules/nextjs-no-edge-og-runtime.rs
use oxc_ast::{
    ast::{BindingPattern, Declaration, Expression},
    AstKind,
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{
    context::{ContextHost, LintContext},
    rule::Rule,
    AstNode,
};

const MESSAGE: &str = "Edge runtime limits OG image generation. Node.js runtime supports more fonts, filesystem access, and larger response sizes.";

#[derive(Debug, Default, Clone)]
pub struct NextjsNoEdgeOgRuntime;

declare_oxc_lint!(
    /// Disallow the Edge runtime in Next.js social image routes.
    NextjsNoEdgeOgRuntime,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Disallow the Edge runtime in social image routes.",
);

impl Rule for NextjsNoEdgeOgRuntime {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        !is_non_production_file(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::ExportDeclaration(export_declaration) = node.kind() else {
            return;
        };
        let filename = ctx.file_path().to_string_lossy().replace('\\', "/");
        if !is_og_image_filename(&filename) || !is_next_file_active(ctx) {
            return;
        }
        let Declaration::VariableDeclaration(variable_declaration) =
            &export_declaration.declaration
        else {
            return;
        };
        let has_edge_runtime = variable_declaration.declarations.iter().any(|declarator| {
            matches!(&declarator.id, BindingPattern::BindingIdentifier(identifier) if identifier.name == "runtime")
                && matches!(&declarator.init, Some(Expression::StringLiteral(value)) if value.value == "edge")
        });
        if has_edge_runtime {
            ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(export_declaration.span));
        }
    }
}

fn is_og_image_filename(filename: &str) -> bool {
    let Some(file_name) = filename.rsplit('/').next() else {
        return false;
    };
    let Some((stem, extension)) = file_name.rsplit_once('.') else {
        return false;
    };
    if !matches!(extension, "ts" | "tsx" | "js" | "jsx" | "mts" | "mjs") {
        return false;
    }
    ["opengraph-image", "twitter-image"].iter().any(|prefix| {
        stem.strip_prefix(prefix)
            .is_some_and(|suffix| suffix.chars().all(|character| character.is_ascii_digit()))
    })
}
